x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineBinaryComparisons.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules;
+
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.predicate.Predicates;
+import org.elasticsearch.xpack.esql.core.expression.predicate.logical.And;
+import org.elasticsearch.xpack.esql.core.expression.predicate.logical.BinaryLogic;
+import org.elasticsearch.xpack.esql.core.expression.predicate.logical.Or;
+import org.elasticsearch.xpack.esql.core.expression.predicate.operator.comparison.BinaryComparison;
+import org.elasticsearch.xpack.esql.core.util.CollectionUtils;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.GreaterThan;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.GreaterThanOrEqual;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.LessThan;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.LessThanOrEqual;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.NotEquals;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class CombineBinaryComparisons extends OptimizerRules.OptimizerExpressionRule<BinaryLogic> {
+
+    public CombineBinaryComparisons() {
+        super(OptimizerRules.TransformDirection.DOWN);
+    }
+
+    @Override
+    public Expression rule(BinaryLogic e) {
+        if (e instanceof And and) {
+            return combine(and);
+        } else if (e instanceof Or or) {
+            return combine(or);
+        }
+        return e;
+    }
+
+    // combine conjunction
+    private static Expression combine(And and) {
+        List<BinaryComparison> bcs = new ArrayList<>();
+        List<Expression> exps = new ArrayList<>();
+        boolean changed = false;
+        List<Expression> andExps = Predicates.splitAnd(and);
+        // Ranges need to show up before BinaryComparisons in list, to allow the latter be optimized away into a Range, if possible.
```

Comment:
I know this is a c&p issue, but reading just the comment hints that Ranges need special handling here, when in fact they're just ignored in theory, but in practice they don't occur at all. I don't think it's worth another CI round, but maybe we can either add a TODO or remove the comment.